---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "",
  fig.path = "README/README-",
  fig.align = 'center',
  fig.retina=2, fig.width = 7, fig.height = 5,
  warning = FALSE,
  message = FALSE
)
options("width"=200)
```

iNEXT.3D (R package)
=====

<h5 align="right">Latest version: `r Sys.Date()`</h5>

<font color="394CAE"><h3 color= 394CAE style = "font-weight: bold"> Introduction to iNEXT.3D (R package): Excerpt from iNEXT.3D User’s Guide </h3> </font>
<br>
<h5><b>Anne Chao, Kai-Hsiang Hu</b>
<br><br>
<i>Institute of Statistics, National Tsing Hua University, Hsin-Chu, Taiwan 30043</i>  
</h5>
<br>
iNEXT.3D (INterpolation and EXTrapolation for three dimensions of diversity) is an R package, available in [Github](https://github.com/AnneChao), for rarefaction and extrapolation of species diversity (Hill numbers) for three dimensions. Here we provide a quick introduction demonstrating how to run iNEXT.3D, and showing three types of sampling curves. See Chao et al. (2021) for methodologies. An online version of [iNEXT.3D Online](https://chao.shinyapps.io/iNEXT_3D/) is also available for users without an R background. Detailed information about all functions in iNEXT.3D is provided in the iNEXT.3D Manual in [iNEXT.3D_vignettes](http://chao.stat.nthu.edu.tw/wordpress/wp-content/uploads/software/A%20Quick%20Introduction%20to%20iNEXT.3D%20via%20Examples.html),which is available from [Anne Chao's website](http://chao.stat.nthu.edu.tw/wordpress/software_download/).

`iNEXT3D` focuses on three measures of Hill numbers of order q: species richness (`q = 0`), Shannon diversity (`q = 1`, the exponential of Shannon entropy) and Simpson diversity (`q = 2`, the inverse of Simpson concentration) among three classes: taxonomic diversity ('TD'), phylogenetic diversity ('PD'), functional diversity ('FD'). For each diversity measure, `iNEXT.3D` uses the observed sample of abundance or incidence data (called the “reference sample”) to compute diversity estimates and the associated confidence intervals for the following two types of rarefaction and extrapolation (R/E):

1. Sample‐size‐based R/E sampling curves:`iNEXT3D` computes diversity estimates for rarefied and extrapolated samples up to an appropriate size. This type of sampling curve plots the diversity estimates with respect to sample size.     
2. Coverage‐based R/E sampling curves: `iNEXT3D` computes diversity estimates for rarefied and extrapolated samples with sample completeness (as measured by sample coverage) up to an appropriate coverage. This type of sampling curve plots the diversity estimates with respect to sample coverage. 

`iNEXT3D` also plots the above two types of sampling curves and a sample completeness curve. The sample completeness curve provides a bridge between these two types of curves.

### SOFTWARE NEEDED TO RUN INEXT.3D IN R

-   Required: [R](http://cran.rstudio.com/)
-   Suggested: [RStudio IDE](http://www.rstudio.com/ide/download/)

### HOW TO RUN INEXT.3D:

The iNEXT.3D package is available on [Github](https://github.com/AnneChao/iNEXT.3D) and can be downloaded with a standard installation procedure using the commands shown below.For a first-time installation, an additional visualization extension package (ggplot2) must be installed and loaded.  

```{r install, eval=FALSE}
## install the latest version from github
install.packages('devtools')
library(devtools)
# install_github('AnneChao/iNEXT.3D')
## import packages
library(iNEXT.3D)
library(ggplot2)
```


In this document, we provide a quick introduction demonstrating how to run the package `iNEXT.3D`(iNterpolation and EXTrapolation in three Dimensions). `iNEXT.3D` has several main function: `iNEXT3D`, `ggiNEXT3D`, `AO3D`, `ggAO3D`, `estimate3D`, and `DataInfo3D.`


### MAIN FUNCTION: iNEXT3D()

The main function iNEXT3D() with default arguments is described below:
<br><br>
iNEXT3D(data, diversity = 'TD', q = c(0,1,2), datatype = "abundance", 
        size = NULL, endpoint = NULL, knots = 40, nboot = 50, conf = 0.95, nT = NULL, 
        PDtree = NULL, PDreftime = NULL, PDtype = 'meanPD', 
        FDdistM, FDtype = 'AUC', FDtau = NULL)
<br><br>
This main function computes diversity estimates of order q, the sample coverage estimates and related statistics for K (if knots=K) evenly-spaced knots (sample sizes) between size 1 and the endpoint, where the endpoint is as described below. Each knot represents a particular sample size for which diversity estimates will be calculated. By default, endpoint is set to be double the reference sample size. 



This function returns an "iNEXT3D" object which can be further used to make plots using the function ggiNEXT3D() to be described below.

```{r, echo=FALSE,warning=FALSE}
Des <- c("data"," (a). For datatype = 'abundance', data can be input as a vector of species abundances (for a single assemblage), matrix/data.frame (species by assemblages), or a list of species abundance vectors. 
         (b). For datatype = 'incidence_freq', data can be input as a vector of incidence frequencies (for a single assemblage), matrix/data.frame (species by assemblages), or a list of incidence frequencies; the first entry in all types of input must be the number of sampling units in each assemblage. 
         (c). For datatype = 'incidence_raw', data can be input as a list of matrix/data.frame (species by sampling units); data can also be input as a matrix/data.frame by merging all sampling units across assemblages based on species identity; in this case, the number of sampling units (nT, see below) must be input.",
"diversity"," selection of diversity type: 'TD' = Taxonomic diversity, 'PD' = Phylogenetic diversity, and 'FD' = Functional diversity.",
"q","	a numerical vector specifying the diversity orders. Default is c(0, 1, 2).",
"datatype"	,"data type of input data: individual-based abundance data (datatype = 'abundance'), sampling-unit-based incidence frequencies data (datatype = 'incidence_freq'), or species by sampling-units incidence matrix (datatype = 'incidence_raw') with all entries being 0 (non-detection) or 1 (detection).", 
"size","an integer vector of sample sizes for which diversity estimates will be computed. If NULL, then diversity estimates will be calculated for those sample sizes determined by the specified/default endpoint and knots;",
"endpoint","an integer specifying the sample size that is the endpoint for R/E calculation; If NULL, then endpoint=double the reference sample size;",
"knots","	an integer specifying the number of equally-spaced knots (say K, default is 40) between size 1 and the endpoint;each knot represents a particular sample size for which diversity estimate will be calculated. If the endpoint is smaller than the reference sample size, then iNEXT3D() computes only the rarefaction esimates for approximately K evenly spaced knots. If the endpoint is larger than the reference sample size, then iNEXT3D() computes rarefaction estimates for approximately K/2 evenly spaced knots between sample size 1 and the reference sample size, and computes extrapolation estimates for approximately K/2 evenly spaced knots between the reference sample size and the endpoint.",
"nboot"," a positive integer specifying the number of bootstrap replications when assessing sampling uncertainty and constructing confidence intervals. Enter 0 to skip the bootstrap procedures. Default is 50.",
"conf","	a positive integer specifying the number of bootstrap replications when assessing sampling uncertainty and constructing confidence intervals. Enter 0 to skip the bootstrap procedures. Default is 50.",
"nT","(required only when datatype = 'incidence_raw' and input data is matrix/data.frame) a vector of nonnegative integers specifying the number of sampling units in each assemblage. If assemblage names are not specified, then assemblages are automatically named as 'assemblage1', 'assemblage2',..., etc.",
"PDtree","	(required only when diversity = 'PD'), a phylogenetic tree in Newick format for all observed species in the pooled assemblage.",
"PDreftime", "(required only when diversity = 'PD'), a vector of numerical values specifying reference times for PD. Default is NULL (i.e., the age of the root of PDtree).",
"PDtype"," (required only when diversity = 'PD'), select PD type: PDtype = 'PD' (effective total branch length) or PDtype = 'meanPD' (effective number of equally divergent lineages). Default is 'meanPD', where meanPD = PD/tree depth.",
"FDdistM"," (required only when diversity = 'FD'), select FD type: FDtype = 'tau_values' for FD under specified threshold values, or FDtype = 'AUC' (area under the curve of tau-profile) for an overall FD which integrates all threshold values between zero and one. Default is 'AUC'.",
"FDtype"," (required only when diversity = 'FD'), select FD type: FDtype = 'tau_values' for FD under specified threshold values, or FDtype = 'AUC' (area under the curve of tau-profile) for an overall FD which integrates all threshold values between zero and one. Default is 'AUC'.",
"FDtau"," (required only when diversity = 'FD' and FDtype = 'tau_values'), a numerical vector between 0 and 1 specifying tau values (threshold levels). If NULL (default), then threshold is set to be the mean distance between any two individuals randomly selected from the pooled assemblage (i.e., quadratic entropy)."
)
output <- 
  matrix(Des, 
         ncol=2, byrow = TRUE)

library(htmlTable)
htmlTable(output,
          header =  c("Argument","Description"),align = "l"
          )
```
This function returns an "iNEXT3D" object which can be further used to make plots using the function ggiNEXT3D() to be described below.

### DATA FORMAT/INFORMATION

Three types of data are supported: ("abundance", "incidence_raw", or "incidence_freq"): 

1.	Individual‐based abundance data (datatype="abundance"): Input data for each assemblage/site include samples species abundances in an empirical sample of n individuals (“reference sample”). When there are N assemblages, input data consist of an S by N abundance matrix, or N lists of species abundances.

2.	Sampling-unit-based incidence data:  There are two kinds of input data. 

(2a) Incidence‐raw data (datatype="incidence_raw"): for each assemblage, input data for a reference sample consist of a species‐by‐sampling‐unit matrix; when there are N assemblages, input data consist of N lists of matrices, and each matrix is a species‐by‐sampling‐unit matrix. If the phylogenetic diversity required, the matrix of combined assemblage is allowed, but nT must be specified (see above description).

(2b) Incidence‐frequency data (datatype="incidence_freq"): input data for each assemblage consist of species sample incidence frequencies (row sums of each incidence matrix). When there are N assemblages, input data consist of an S by N matrix, or N lists of species incidence frequencies. The first entry of each list must be the total number of sampling units, followed by the species incidence frequencies.

Two data sets (beetles for abundance data ) are included in iNEXT3D package. The first list of beetles’ data consist of abundance data who live on three treatments (“Control”, “Debarked” and “Scratched”) of trees. The second list consist of the pylogenetic tree for every beetles. And the third list consist of distance matrix for each pair of beetles. For these data, the following commands display the sample species abundances and run the iNEXT3D() function for three types of diversty ("TD", "PD", "FD" with threshold dmean, "AUC" which integate FD from threshold 0-1) in q = 0.

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(comment="", message=FALSE, warning=FALSE)
library(iNEXT.3D)
library(ggplot2)
```

```{r}
data("dunes")

out.TD <- iNEXT3D(data = dunes$data, diversity = "TD", 
               q = c(0, 1, 2), datatype = "abundance", 
               nboot = 10)
out.PD <- iNEXT3D(data = dunes$data, diversity = "PD", 
               q = c(0, 1, 2), datatype = "abundance", 
               PDtree = dunes$tree, 
               nboot = 10)
out.FD <- iNEXT3D(data = dunes$data, diversity = "FD", 
               q = c(0, 1, 2), datatype = "abundance", 
               FDdistM = dunes$dist,
               nboot = 5)

```

```{r}
out.TD
```

The second data frame of list $iNextEst (as shown below for ‘coverage_based’) under each class includes the reference sample coverage (‘goalSC’), real sample coverage (‘SC’), sample size (m, i.e., each of the 40 knots), the method (Rarefaction, Observed, or Extrapolation, depending on whether the size m is less than, equal to, or greater than the reference sample size), the diversity order, the diversity estimate of order q conditioning on ‘goalSC’, the 95% lower and upper confidence limits of diversity. These sample coverage estimates with confidence intervals are used for plotting the coverage-based R/E curves.

$AsyEst lists the observed diversity, asymptotic estimates, estimated bootstrap s.e. and 95% confidence intervals for Hill numbers with q = 0, 1, and 2. The estimated asymptotes and the observed diversity are calculated via the functions AO3D() . The output for the dunes data is shown below. All row and column variables are self‐explanatory.


### BASIC GRAPHIC DISPLAYS: FUNCTION ggiNEXT3D()
The ggiNEXT3D() function, which extends ggplot2 to the "iNEXT3D" object, is described as follows with default arguments:
<br><br>
ggiNEXT3D(outcome, type = 1:3, se = TRUE,
          facet.var = "Assemblage", color.var = "Order.q")  
<br><br>
Here outcome is the object of iNEXT3D()’s output. Three types of curves are allowed for different diversity classes:

(a) Sample-size-based R/E curve (type=1): see Figs. 1a and 2a in the main text. This curve plots diversity estimates with confidence intervals (if se=TRUE) as a function of sample size up to double the reference sample size, by default, or a user‐specified endpoint.

(b) Sample completeness curve (type=2) with confidence intervals (if se=TRUE): see Figs. 1b and 2b in the main text. This curve plots the sample coverage with respect to sample size for the same range described in (1).

(c) Coverage-based R/E curve (type=3): see Figs. 1c and 2c in the main text. This curve plots the diversity estimates with confidence intervals (if se=TRUE) as a function of sample coverage up to the maximum coverage obtained from the maximum size described in (1).

<br><br>
The argument facet.var=("Order.q", "Assemblage") is used to create a separate plot for each value of the specified variable. For example, the following code displays a separate plot (in Figs 1a and 1c) for each value of the diversity order q. The ggiNEXT3D() function is a wrapper around ggplot2 package to create a R/E curve using a single line of code. The resulting object is of class "ggplot", so can be manipulated using the ggplot2 tools.

The argument facet.var="Assemblage" in ggiNEXT3D function creates a separate plot for each assembalge, therefore the different Order.q will seperated by different colours as shown below:


### RAREFACTION/EXTRAPOLATION FOR ABUNDANCE DATA 

```{r}
# Sample‐size‐based R/E curves, separating by "assemblage""
ggiNEXT3D(out.TD, type = 1, facet.var = "Assemblage")
```

The argument facet.var="Order.q" in ggiNEXT3D function creates a separate plot for each order, therefore three assemblages will be seperated by different colours as shown below:

```{r}
# Sample-size-based R/E curves, separating plots by "Order.q"
ggiNEXT3D(out.TD, type = 1, facet.var = "Order.q")
```

The following commands return the sample completeness curve in which different colors are used for the three assemblages. Since the sample completeness curve are same for differnet class of diversity, ggiNEXT3D returns only one plot:

```{r}
ggiNEXT3D(out.TD, type = 2, facet.var = "Order.q", color.var="Assemblage")
```

The following commands return the coverage‐based R/E sampling curves in which different colors are used for the three assemblages (facet.var="Assemblage") and for three orders (facet.var="Order.q")

```{r}
ggiNEXT3D(out.TD, type = 3, facet.var="Assemblage")
```

```{r}
ggiNEXT3D(out.TD, type = 3, facet.var="Order.q")
```


### RAREFACTION/EXTRAPOLATION FOR RAW INCIDENCE DATA: (incidence_raw)

For illustration, we use the Hinkley’s fish data (in the dataset fish included in the package) at three time periods (1981-1985, 1987-1991 and 2015-2019). Incidence raw data is allowed for all diversity class.This data set (fish) included in the package is three list of matrices; each matrix is a species by plots data.frame for a time period. We only use Taxonomic diversity(TD) for demonstration below.


```{r}
data("fish")
head(fish[[1]]$`1981-1985`,4)
```

```{r}
# fish.tree = fish$tree
# fish.dis = fish$dist

out.raw <- iNEXT3D(data = fish$data, 
                   diversity = "TD",
                   q = c(0, 1, 2), datatype = "incidence_raw",nboot = 10)
ggiNEXT3D(out.raw, type = 1)
```

```{r}
ggiNEXT3D(out.raw, type = 2)
```

```{r}
ggiNEXT3D(out.raw, type = 3)
```


### DATA INFORMATION FUNCTION: DataInfo3D()

We can supply the function
<br><br>
DataInfo3D(data, diversity = "TD", datatype = "abundance", 
  nT = NULL, PDtree, PDreftime = NULL, 
  FDdistM, FDtype = "AUC", FDtau = NULL) 
<br><br>
to compute three type diversity(‘TD’,‘PD’,‘FD’) data information, which including sample size, observed species richness, sample coverage estimate, and the first ten abundance/incidence frequency counts, and so on.

```{r}
DataInfo3D(dunes$data, diversity = 'TD', datatype = "abundance")
```



### POINT ESTIMATION FUNCTION: estimate3D()

We also supply the function
<br><br>
estimate3D(data, diversity = "TD", q = c(0, 1, 2), datatype = "abundance", 
  base = "coverage", level = NULL, nboot = 50, conf = 0.95, 
  nT = NULL, PDtree, PDreftime = NULL, PDtype = "meanPD", 
  FDdistM, FDtype = "AUC", FDtau = NULL)
<br><br>
to compute three type diversity(‘TD’,‘PD’,‘FD’) estimates with q = 0, 1, 2 for any particular level of sample size (base="size") or any specified level of sample coverage (base="coverage") for either abundance data (datatype="abundance") or incidence data (datatype="incidence_freq" or "incidence_raw"). If level=NULL, this function computes the diversity estimates for the minimum sample size/coverage among all assemblages.

For example, the following command returns the taxonomic diversity (‘TD’) with a specified level of sample coverage of 93% for the dunes data. For some assemblages, this coverage value corresponds to the rarefaction part whereas the others correspond to extrapolation, as indicated in the method of the output.

```{r}
estimate3D(dunes$data, diversity = 'TD', q = c(0,1,2), datatype = "abundance", base = "coverage",level = 0.93)
```



### EMPIRICAL AND ASYMPTOTIC DIVERSITY FUNCTION: AO3D

<br><br>
AO3D(
  data,diversity = "TD",q = seq(0,2,0.2),datatype = "abundance",
  nboot = 50,conf = 0.95,nT = NULL,method = c("Asymptotic", "Observed"),
  PDtree,PDreftime = NULL,PDtype = "meanPD",
  FDdistM,FDtype = "AUC",FDtau = NULL
)
<br><br>

The function AO3D() can compute three type diversity(‘TD’,‘PD’,‘FD’),which including empirical diversity and asymptotic diversity. For either abundance data (datatype="abundance") or incidence data(datatype="incidence_freq" or "incidence_raw")with any specified level of q can be compute.

For example, the following command returns an empirical taxonomic diversity(‘TD’) and asymptotic taxonomic diversity(‘TD’) for dunes data, along with its confidence interval, for a specified q level from 0 to 2.

```{r}
out1 <- AO3D(dunes$data, diversity = 'TD', datatype = "abundance", method = c("Asymptotic", "Observed"),nboot = 5,conf = 0.95)

out1
```


### GRAPHIC DISPLAYS FUNCTION: ggAO3D()

Plots q-profile, time-profile, and tau-profile based on the outcome of AO3D using the ggplot2 package.

The function ggAO3D(), which extends ggplot2 with default arguments, is described as follows:
<br><br>
ggAO3D(outcome, profile = "q")
<br><br>
Here outcome is the object of AO3D’s output and profile is an object versus to diversity.Default is profile = "q".Note that profile = "time" is allowed for only when diversity = “PD” and profile = "tau" profile is allowed for only when diversity = "FD" and FDtype = "tau_values".

```{r}
# q profile curve""
ggAO3D(out1,profile = "q")
```

The argument profile = "time" in ggAO3D function creates a separate plot for each order.q. Therefore the different assemblages will seperated by different colours as shown below:


```{r}
# time curves, separating by "Order.q"
data(data.inc)
data <- data.inc$data
tree <- data.inc$tree
nT <- data.inc$nT
out2 <- AO3D(data, diversity = 'PD', q = c(0, 1, 2), datatype = "incidence_raw", nT = nT, nboot = 5, method = c("Asymptotic", "Observed"), PDtree = tree, PDreftime = seq(0.1, 82.8575, length.out = 40))
ggAO3D(out2, profile = "time")
```

The argument profile = "tau" in ggAO3D function creates a separate plot for each order.q. Therefore the different assemblages will seperated by different colours as shown below:

```{r}
# tau curves, separating by "Order.q"
data <- dunes$data
distM <-  dunes$dist
out3 <- AO3D(data, diversity = 'FD', q = c(0, 1, 2), datatype = "abundance", nboot = 5, method = c("Asymptotic", "Observed"), FDtau = seq(0, 1, 0.1), FDdistM = distM, FDtype = 'tau_values')
ggAO3D(out3, profile = "tau")
```



### How to cite
- Chao, A., Henderson, P. A., Chiu, C.-H., Moyes, F., Hu, K.-H., Dornelas, M and Magurran, A. E. (2021). Measuring temporal change in alpha diversity: a framework integrating taxonomic, phylogenetic and functional diversity and the iNEXT.3D standardization. Methods in Ecology and Evolution, 12, 1926-1940.


### Referance

Chao, A., Henderson, P. A., Chiu, C.-H., Moyes, F., Hu, K.-H., Dornelas, M and Magurran, A. E. (2021). Measuring temporal change in alpha diversity: a framework integrating taxonomic, phylogenetic and functional diversity and the iNEXT.3D standardization. Methods in Ecology and Evolution, 12, 1926-1940.

